blaze-client/src/main/scala/org/http4s/blaze/client/Http1Connection.scala

```diff
@@ -217,49 +205,49 @@ private final class Http1Connection[F[_]](
-            val idleTimeoutS = timeoutFiber.joinWithNever.attempt.map {
-              case Right(t) => Left(t): Either[Throwable, Unit]
-              case Left(t) => Left(t): Either[Throwable, Unit]
-            }
-
-            val writeRequest: F[Boolean] = getChunkEncoder(req, mustClose, rr)
-              .write(rr, req.body)
-              .guarantee(F.delay(resetWrite()))
-              .onError {
-                case EOF => F.unit
-                case t => F.delay(logger.error(t)("Error rendering request"))
-              }
-
-            val response: F[Resource[F, Response[F]]] =
-              F.bracketCase(
-                writeRequest.start
-              )(writeFiber =>
-                receiveResponse(
-                  mustClose,
-                  doesntHaveBody = req.method == Method.HEAD,
-                  idleTimeoutS,
-                  idleRead,
-                  // We need to wait for the write to complete so that by the time we attempt to recycle the connection it is fully idle.
-                ).map(response => Resource.onFinalize(writeFiber.join.attempt.void).as(response))
-              ) {
-                case (_, Outcome.Succeeded(_)) => F.unit
-                case (writeFiber, Outcome.Canceled() | Outcome.Errored(_)) => writeFiber.cancel
-              }
-
-            F.race(response, timeoutFiber.joinWithNever)
+            F.bracketCase(
+              writeRequest.start
+            )(writeFiber =>
+              receiveResponse(
+                mustClose,
+                doesntHaveBody = req.method == Method.HEAD,
+                cancellation.race(timeoutFiber.joinWithNever).map(e => Left(e.merge)),
+                idleRead,
+              ).map(response =>
+                // We need to finish writing before we attempt to recycle the connection. We consider three scenarios:
+                // - The write already finished before we got the response. This is the most common scenario. `join` completes immediately.
+                // - The whole request was already transmitted and we received the response from the server, but we did not yet notice that the write is complete. This is sort of a race, it happens frequently enough when load testing. We need to wait just a moment for the `join` to finish.
+                // - The server decided to reject our request before we finished sending it. The server responded (typically with an error) and closed the connection. We shouldn't wait for the `writeFiber`. This connection needs to be disposed.
+                Resource.make(F.pure(response))(_ =>
+                  writeFiber.join.attempt.race(closed.get >> writeFiber.cancel.start).void
+                )
+              )
+            ) {
+              case (_, Outcome.Succeeded(_)) => F.unit
+              case (_, Outcome.Canceled()) => F.delay(shutdown())
+              case (_, Outcome.Errored(e)) => F.delay(shutdownWithError(e))
+            }.race(timeoutFiber.joinWithNever)
```

Comment:
I think that would be the request or the response header timeout.  I think it would be weird for that to fire down here and not in `receiveResponse`, but I also don't see the harm in merging all of that into one timeout (which we do anyway) and handling them together.  I'll change it.